lib/authenticate_methods.rb

```diff
@@ -74,6 +54,54 @@ def authenticate_admin!
 
   protected #################
 
+  def use_signed_params_and_redirect
+    log_out_unlinked_student || log_in_known_LMS_user || prepare_for_new_LMS_user
+  end
+
+  def known_LMS_user
+    @known_LMS_user ||= UserExternalUuid.find_by_uuid(signed_params['uuid'])&.user
+  end
+
+  # If a student is already logged in, but their account is not yet linked with the LMS, log them out.
+  def log_out_unlinked_student
+    if signed_in? && !known_LMS_user && current_user.student?
+      sign_out!(security_log_data: {reason: 'LMS student'})
+      redirect_to(newflow_signup_student_path(request.query_parameters))
+      return true
+    end
+  end
+
+  def log_in_known_LMS_user
+    if signed_in? && known_LMS_user != current_user
+      sign_out!(security_log_data: {type: 'different external user'})
+    end
+
+    if known_LMS_user
+      sign_in!(known_LMS_user, security_log_data: {type: 'external uuid'})
+      # redirect_to(newflow_login_path(request_url_without_signed_params))
+      return true
+    end
+
+    false
+  end
+
+  def prepare_for_new_LMS_user
+    user = Newflow::FindOrCreateUserFromSignedParams.call(signed_params).outputs.user
+    session[:user_from_signed_params] = user
+
+    if (new_lms_user = session[:user_from_signed_params])
+      if new_lms_user['role'] == 'student' && new_lms_user['state'] != 'activated'
```

Comment:
see comment above for literal strings.  lots and lots of duplication here.

Reply:
consider:  cleaning up this nested IF.   couple of ways.   (1) create a small PORO to do the role logic (and unit test that separately).  or (2) create methods of these if expressions and use a case statement.  

both these ideas would clean this up..each having their own merits.

Reply:
Changed to:

```

  def prepare_for_new_LMS_user
    user = Newflow::FindOrCreateUserFromSignedParams.call(signed_params).outputs.user
    session[:user_from_signed_params] = user

    if (new_lms_user = session[:user_from_signed_params])
      if new_lms_user.student? && !new_lms_user.activated?
        redirect_to newflow_signup_student_path(request.query_parameters)
      elsif new_lms_user.student? && new_lms_user.activated?
        redirect_to newflow_login_path(request.query_parameters)
      elsif new_lms_user.instructor? && new_lms_user.unverified?
        redirect_to educator_signup_path(request.query_parameters)
      else
        redirect_to newflow_login_path(request.query_parameters)
      end
    end
  end
```

Reply:
I would love to learn how I can improve it beyond that. This suggestion, by the way, was EXCELLENT. Thanks!